// @langchain/langgraph-sdk/dist/stream/projections/values.js
import { ensureMessageInstances } from "../../ui/messages.js";
import { isRootNamespace, namespaceKey } from "../namespace.js";
import { openProjectionSubscription } from "./runtime.js";
//#region src/stream/projections/values.ts
function valuesProjection(namespace, messagesKey = "messages") {
	const ns = [...namespace];
	return {
		key: `values|${messagesKey}|${namespaceKey(ns)}`,
		namespace: ns,
		initial: void 0,
		open({ thread, store, rootBus }) {
			const applyValuesEvent = (event) => {
				const coerced = coerceMessagesInState(event.params.data, messagesKey);
				store.setValue(coerced);
			};
			if (isRootNamespace(ns) && rootBus.channels.includes("values")) {
				const unsubscribe = rootBus.subscribe((event) => {
					if (event.method !== "values") return;
					if (!isRootNamespace(event.params.namespace)) return;
					applyValuesEvent(event);
				});
				return { dispose() {
					unsubscribe();
				} };
			}
			return openProjectionSubscription({
				thread,
				channels: ["values"],
				namespace: ns,
				onEvent(event) {
					if (event.method !== "values") return;
					applyValuesEvent(event);
				}
			});
		}
	};
}
function coerceMessagesInState(value, messagesKey) {
	if (value == null || typeof value !== "object" || Array.isArray(value)) return value;
	const state = value;
	const maybeMessages = state[messagesKey];
	if (!Array.isArray(maybeMessages) || maybeMessages.length === 0) return value;
	if (!maybeMessages.some((m) => m != null && typeof m.getType !== "function")) return value;
	return {
		...state,
		[messagesKey]: ensureMessageInstances(maybeMessages)
	};
}
//#endregion
export { valuesProjection };
