// @langchain/langgraph-sdk/dist/client/stream/handles/tools.cjs
//#region src/client/stream/handles/tools.ts
/**
* Incrementally assembles `tools` events into complete
* {@link AssembledToolCall} objects with promise-based output/status/error.
*
* Each `tool-started` event produces an {@link AssembledToolCall} whose
* `output`, `status`, and `error` promises resolve when `tool-finished`
* or `tool-error` arrives for the same `tool_call_id`.
*/
var ToolCallAssembler = class {
	active = /* @__PURE__ */ new Map();
	consume(event) {
		const data = event.params.data;
		if (data.event === "tool-started") return this.handleStarted(event, data);
		if (data.event === "tool-finished") {
			this.handleFinished(data);
			return;
		}
		if (data.event === "tool-error") {
			this.handleError(data);
			return;
		}
	}
	/**
	* Reject any in-flight tool calls (e.g. on session close).
	*/
	failAll(reason) {
		for (const tc of this.active.values()) {
			tc.rejectOutput(reason);
			tc.resolveStatus("error");
			tc.resolveError(reason.message);
		}
		this.active.clear();
	}
	handleStarted(event, data) {
		let resolveOutput;
		let rejectOutput;
		let resolveStatus;
		let resolveError;
		const output = new Promise((resolve, reject) => {
			resolveOutput = resolve;
			rejectOutput = reject;
		});
		output.catch(() => void 0);
		const status = new Promise((resolve) => {
			resolveStatus = resolve;
		});
		const error = new Promise((resolve) => {
			resolveError = resolve;
		});
		const entry = {
			name: data.tool_name,
			callId: data.tool_call_id,
			namespace: [...event.params.namespace],
			input: data.input,
			resolveOutput,
			rejectOutput,
			resolveStatus,
			resolveError
		};
		this.active.set(data.tool_call_id, entry);
		return {
			name: entry.name,
			callId: entry.callId,
			namespace: entry.namespace,
			input: entry.input,
			output,
			status,
			error
		};
	}
	handleFinished(data) {
		const entry = this.active.get(data.tool_call_id);
		if (!entry) return;
		this.active.delete(data.tool_call_id);
		entry.resolveOutput(data.output);
		entry.resolveStatus("finished");
		entry.resolveError(void 0);
	}
	handleError(data) {
		const entry = this.active.get(data.tool_call_id);
		if (!entry) return;
		this.active.delete(data.tool_call_id);
		entry.rejectOutput(new Error(data.message));
		entry.resolveStatus("error");
		entry.resolveError(data.message);
	}
};
//#endregion
exports.ToolCallAssembler = ToolCallAssembler;
