// frontend/src/components/auth/login-form.tsx
'use client';

import React, { useState } from 'react';
import { useRouter } from 'next/navigation';
import { Eye, EyeOff, Loader2 } from 'lucide-react';
import Link from 'next/link';
import { createClient } from '@/lib/supabase/client';
import { Logo } from '@/components/shared/logo';

interface LoginFormState {
  email: string;
  password: string;
  showPassword: boolean;
  isLoading: boolean;
  error: string | null;
}

export function LoginForm(): React.JSX.Element {
  const router = useRouter();
  const supabase = createClient();

  const [state, setState] = useState<LoginFormState>({
    email: '',
    password: '',
    showPassword: false,
    isLoading: false,
    error: null,
  });

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>): Promise<void> => {
    e.preventDefault();
    setState((prev) => ({ ...prev, isLoading: true, error: null }));

    try {
      const { error } = await supabase.auth.signInWithPassword({
        email: state.email,
        password: state.password,
      });

      if (error) {
        setState((prev) => ({ ...prev, error: error.message, isLoading: false }));
        return;
      }

      router.push('/dashboard');
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : 'An error occurred';
      setState((prev) => ({ ...prev, error: errorMessage, isLoading: false }));
    }
  };

  const updateState = (updates: Partial<LoginFormState>): void => {
    setState((prev) => ({ ...prev, ...updates }));
  };

  return (
    <div>
      {/* Logo */}
      <Logo size="md" showText={false} />

      {/* Title */}
      <h1 className="text-3xl font-semibold text-[#0F172A] mt-8">Log in</h1>

      {/* Subtitle */}
      <p className="text-sm text-[#64748B] mt-2">
        Welcome back! Please enter your details.
      </p>

      {/* Form */}
      <form onSubmit={handleSubmit} className="mt-8 space-y-5">
        {/* Email field */}
        <div>
          <label className="label">Email</label>
          <input
            type="email"
            className="input"
            placeholder="Enter your email"
            value={state.email}
            onChange={(e) => updateState({ email: e.target.value })}
            required
            disabled={state.isLoading}
          />
        </div>

        {/* Password field */}
        <div>
          <label className="label">Password</label>
          <div className="relative">
            <input
              type={state.showPassword ? 'text' : 'password'}
              className="input pr-10"
              placeholder="Enter your password"
              value={state.password}
              onChange={(e) => updateState({ password: e.target.value })}
              required
              disabled={state.isLoading}
            />
            <button
              type="button"
              onClick={() => updateState({ showPassword: !state.showPassword })}
              className="absolute right-3 top-1/2 -translate-y-1/2 text-[#94A3B8] hover:text-[#64748B] transition-colors"
              disabled={state.isLoading}
            >
              {state.showPassword ? (
                <EyeOff size={16} />
              ) : (
                <Eye size={16} />
              )}
            </button>
          </div>
        </div>

        {/* Error alert */}
        {state.error && (
          <div className="bg-[#FEE2E2] border border-[#FECACA] rounded-lg px-4 py-3 text-sm text-[#DC2626]">
            {state.error}
          </div>
        )}

        {/* Submit button */}
        <button
          type="submit"
          className="btn-primary w-full"
          disabled={state.isLoading}
        >
          {state.isLoading ? (
            <>
              <Loader2 size={16} className="animate-spin" />
              Signing in...
            </>
          ) : (
            'Login'
          )}
        </button>
      </form>

      {/* Footer */}
      <p className="mt-6 text-center text-sm text-[#64748B]">
        Don&apos;t have an account?{' '}
        <Link
          href="/signup"
          className="text-[#2563EB] font-medium hover:underline"
        >
          Sign up
        </Link>
      </p>
    </div>
  );
}
