// frontend/src/components/auth/signup-form.tsx
'use client';

import React, { useState } from 'react';
import { Eye, EyeOff, Loader2, Mail } from 'lucide-react';
import Link from 'next/link';
import { createClient } from '@/lib/supabase/client';
import { Logo } from '@/components/shared/logo';

interface SignupFormState {
  fullName: string;
  email: string;
  password: string;
  confirmPassword: string;
  showPassword: boolean;
  isLoading: boolean;
  error: string | null;
  success: boolean;
}

export function SignupForm(): React.JSX.Element {
  const supabase = createClient();

  const [state, setState] = useState<SignupFormState>({
    fullName: '',
    email: '',
    password: '',
    confirmPassword: '',
    showPassword: false,
    isLoading: false,
    error: null,
    success: false,
  });

  const handleSubmit = async (e: React.FormEvent<HTMLFormElement>): Promise<void> => {
    e.preventDefault();

    // Validation
    if (state.password.length < 8) {
      setState((prev) => ({
        ...prev,
        error: 'Password must be at least 8 characters',
      }));
      return;
    }

    if (state.password !== state.confirmPassword) {
      setState((prev) => ({
        ...prev,
        error: 'Passwords do not match',
      }));
      return;
    }

    setState((prev) => ({ ...prev, isLoading: true, error: null }));

    try {
      const { error } = await supabase.auth.signUp({
        email: state.email,
        password: state.password,
        options: {
          data: {
            full_name: state.fullName,
          },
        },
      });

      if (error) {
        setState((prev) => ({ ...prev, error: error.message, isLoading: false }));
        return;
      }

      setState((prev) => ({ ...prev, success: true, isLoading: false }));
    } catch (err) {
      const errorMessage = err instanceof Error ? err.message : 'An error occurred';
      setState((prev) => ({ ...prev, error: errorMessage, isLoading: false }));
    }
  };

  const updateState = (updates: Partial<SignupFormState>): void => {
    setState((prev) => ({ ...prev, ...updates }));
  };

  // Success state
  if (state.success) {
    return (
      <div className="text-center">
        <Logo size="md" showText={false} />

        <div className="w-16 h-16 bg-[#DCFCE7] rounded-full flex items-center justify-center mx-auto mt-8">
          <Mail size={28} className="text-[#16A34A]" />
        </div>

        <h3 className="text-lg font-semibold text-[#0F172A] mt-4">
          Check your email
        </h3>

        <p className="text-sm text-[#64748B] mt-2">
          We sent a confirmation link to <span className="font-medium">{state.email}</span>
        </p>

        <Link href="/login" className="btn-primary mt-6 inline-flex">
          Back to login
        </Link>
      </div>
    );
  }

  return (
    <div>
      {/* Logo */}
      <Logo size="md" showText={false} />

      {/* Title */}
      <h1 className="text-3xl font-semibold text-[#0F172A] mt-8">Sign up</h1>

      {/* Subtitle */}
      <p className="text-sm text-[#64748B] mt-2">
        Please enter your details.
      </p>

      {/* Form */}
      <form onSubmit={handleSubmit} className="mt-8 space-y-5">
        {/* Full Name field */}
        <div>
          <label className="label">Full Name</label>
          <input
            type="text"
            className="input"
            placeholder="Enter your full name"
            value={state.fullName}
            onChange={(e) => updateState({ fullName: e.target.value })}
            required
            disabled={state.isLoading}
          />
        </div>

        {/* Email field */}
        <div>
          <label className="label">Email</label>
          <input
            type="email"
            className="input"
            placeholder="Enter your email"
            value={state.email}
            onChange={(e) => updateState({ email: e.target.value })}
            required
            disabled={state.isLoading}
          />
        </div>

        {/* Password field */}
        <div>
          <label className="label">Password</label>
          <div className="relative">
            <input
              type={state.showPassword ? 'text' : 'password'}
              className="input pr-10"
              placeholder="Enter your password"
              value={state.password}
              onChange={(e) => updateState({ password: e.target.value })}
              required
              disabled={state.isLoading}
            />
            <button
              type="button"
              onClick={() => updateState({ showPassword: !state.showPassword })}
              className="absolute right-3 top-1/2 -translate-y-1/2 text-[#94A3B8] hover:text-[#64748B] transition-colors"
              disabled={state.isLoading}
            >
              {state.showPassword ? (
                <EyeOff size={16} />
              ) : (
                <Eye size={16} />
              )}
            </button>
          </div>
        </div>

        {/* Confirm Password field */}
        <div>
          <label className="label">Confirm Password</label>
          <div className="relative">
            <input
              type={state.showPassword ? 'text' : 'password'}
              className="input pr-10"
              placeholder="Confirm your password"
              value={state.confirmPassword}
              onChange={(e) => updateState({ confirmPassword: e.target.value })}
              required
              disabled={state.isLoading}
            />
            <button
              type="button"
              onClick={() => updateState({ showPassword: !state.showPassword })}
              className="absolute right-3 top-1/2 -translate-y-1/2 text-[#94A3B8] hover:text-[#64748B] transition-colors"
              disabled={state.isLoading}
            >
              {state.showPassword ? (
                <EyeOff size={16} />
              ) : (
                <Eye size={16} />
              )}
            </button>
          </div>
        </div>

        {/* Error alert */}
        {state.error && (
          <div className="bg-[#FEE2E2] border border-[#FECACA] rounded-lg px-4 py-3 text-sm text-[#DC2626]">
            {state.error}
          </div>
        )}

        {/* Submit button */}
        <button
          type="submit"
          className="btn-primary w-full"
          disabled={state.isLoading}
        >
          {state.isLoading ? (
            <>
              <Loader2 size={16} className="animate-spin" />
              Creating account...
            </>
          ) : (
            'Sign up'
          )}
        </button>
      </form>

      {/* Footer */}
      <p className="mt-6 text-center text-sm text-[#64748B]">
        Already have an account?{' '}
        <Link
          href="/login"
          className="text-[#2563EB] font-medium hover:underline"
        >
          Login
        </Link>
      </p>
    </div>
  );
}
