// @langchain/core/dist/utils/uuid/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_runtime = require("../../_virtual/_rolldown/runtime.cjs");
const require_max = require("./max.cjs");
const require_nil = require("./nil.cjs");
const require_validate = require("./validate.cjs");
const require_parse = require("./parse.cjs");
const require_stringify = require("./stringify.cjs");
const require_v1 = require("./v1.cjs");
const require_v4 = require("./v4.cjs");
const require_v5 = require("./v5.cjs");
const require_v7 = require("./v7.cjs");
const require_version = require("./version.cjs");
//#region src/utils/uuid/index.ts
var uuid_exports = /* @__PURE__ */ require_runtime.__exportAll({
	MAX: () => require_max.default,
	NIL: () => require_nil.default,
	parse: () => require_parse.default,
	stringify: () => require_stringify.default,
	v1: () => require_v1.default,
	v4: () => require_v4.default,
	v5: () => require_v5.default,
	v7: () => require_v7.default,
	validate: () => require_validate.default,
	version: () => require_version.default
});
//#endregion
exports.MAX = require_max;
exports.NIL = require_nil;
exports.parse = require_parse;
exports.stringify = require_stringify;
Object.defineProperty(exports, "uuid_exports", {
	enumerable: true,
	get: function() {
		return uuid_exports;
	}
});
exports.v1 = require_v1;
exports.v4 = require_v4;
exports.v5 = require_v5;
exports.v7 = require_v7;
exports.validate = require_validate;
exports.version = require_version;
