// @langchain/langgraph-sdk/dist/client/stream/transport/agent-server.js
import { ProtocolSseTransportAdapter } from "./http.js";
import { ProtocolWebSocketTransportAdapter } from "./websocket.js";
//#region src/client/stream/transport/agent-server.ts
var HttpAgentServerAdapter = class {
	threadId;
	#delegate;
	constructor(options) {
		this.threadId = options.threadId;
		this.#delegate = options.webSocketFactory != null ? new ProtocolWebSocketTransportAdapter({
			apiUrl: options.apiUrl,
			threadId: options.threadId,
			defaultHeaders: options.defaultHeaders,
			onRequest: options.onRequest,
			paths: options.paths,
			webSocketFactory: options.webSocketFactory
		}) : new ProtocolSseTransportAdapter({
			apiUrl: options.apiUrl,
			threadId: options.threadId,
			defaultHeaders: options.defaultHeaders,
			onRequest: options.onRequest,
			fetch: options.fetch,
			paths: options.paths
		});
	}
	open() {
		return this.#delegate.open();
	}
	send(command) {
		return this.#delegate.send(command);
	}
	events() {
		return this.#delegate.events();
	}
	openEventStream(params) {
		if (this.#delegate.openEventStream == null) throw new Error("HttpAgentServerAdapter delegate does not support openEventStream (WebSocket path).");
		return this.#delegate.openEventStream(params);
	}
	close() {
		return this.#delegate.close();
	}
};
//#endregion
export { HttpAgentServerAdapter };
