// @langchain/langgraph-sdk/dist/client/stream/transport/http.cjs
const require_queue = require("./queue.cjs");
const require_utils = require("./utils.cjs");
const require_decoder = require("./decoder.cjs");
const require_stream = require("./stream.cjs");
//#region src/client/stream/transport/http.ts
/**
* Transport adapter that speaks the thread-centric protocol over HTTP
* commands plus SSE event streams. Bound to a specific `threadId`
* at construction. Each {@link openEventStream} call opens an independent
* filtered SSE connection via `POST /threads/:thread_id/stream/events`.
*/
var ProtocolSseTransportAdapter = class {
	threadId;
	queue = new require_queue.AsyncQueue();
	fetchImpl;
	apiUrl;
	defaultHeaders;
	onRequest;
	fetchFactory;
	commandsUrl;
	streamUrl;
	sessionAbortController = new AbortController();
	eventStreams = /* @__PURE__ */ new Set();
	closed = false;
	constructor(options) {
		this.fetchImpl = options.fetch ?? fetch;
		this.apiUrl = options.apiUrl;
		this.defaultHeaders = options.defaultHeaders ?? {};
		this.onRequest = options.onRequest;
		this.fetchFactory = options.fetchFactory;
		this.threadId = options.threadId;
		this.commandsUrl = options.paths?.commands ?? `/threads/${this.threadId}/commands`;
		this.streamUrl = options.paths?.stream ?? `/threads/${this.threadId}/stream/events`;
	}
	async resolveFetch() {
		if (this.fetchFactory) return await this.fetchFactory();
		return this.fetchImpl;
	}
	/**
	* HTTP/SSE transports have no handshake — connections are made
	* per-command and per-subscription.
	*/
	async open() {}
	async send(command) {
		const response = await this.request(this.commandsUrl, {
			method: "POST",
			headers: { "content-type": "application/json" },
			body: JSON.stringify(command),
			signal: this.sessionAbortController.signal
		});
		if (response.status === 202 || response.status === 204) return;
		const payload = await response.json();
		if (!require_utils.isProtocolResponse(payload)) throw new Error("Protocol command did not return a valid response.");
		return payload;
	}
	/**
	* WebSocket-style single event stream.
	* For the SSE transport this returns a dummy iterable; real event
	* delivery happens via {@link openEventStream}.
	*/
	events() {
		const queue = this.queue;
		return { [Symbol.asyncIterator]: () => ({
			next: async () => await queue.shift(),
			return: async () => {
				queue.close();
				return {
					done: true,
					value: void 0
				};
			}
		}) };
	}
	openEventStream(params) {
		if (this.closed) throw new Error("Protocol transport is closed.");
		const ac = new AbortController();
		this.eventStreams.add(ac);
		const streamQueue = new require_queue.AsyncQueue();
		const streamUrl = this.streamUrl;
		let resolveReady;
		let rejectReady;
		const ready = new Promise((resolve, reject) => {
			resolveReady = resolve;
			rejectReady = reject;
		});
		const since = params.since;
		const startStream = async () => {
			try {
				const response = await this.request(streamUrl, {
					method: "POST",
					headers: {
						"content-type": "application/json",
						accept: "text/event-stream"
					},
					body: JSON.stringify({
						channels: params.channels,
						...params.namespaces ? { namespaces: params.namespaces } : {},
						...params.depth != null ? { depth: params.depth } : {},
						...typeof since === "number" ? { since } : {}
					}),
					signal: ac.signal
				});
				resolveReady();
				const stream = (response.body ?? new ReadableStream({ start(controller) {
					controller.close();
				} })).pipeThrough(require_decoder.BytesLineDecoder()).pipeThrough(require_decoder.SSEDecoder());
				const iterable = require_stream.IterableReadableStream.fromReadableStream(stream);
				for await (const event of iterable) {
					if (ac.signal.aborted || this.closed) break;
					if (require_utils.isRecord(event.data)) {
						const msg = event.data;
						streamQueue.push(msg);
					}
				}
				streamQueue.close();
			} catch (error) {
				rejectReady(error);
				if (ac.signal.aborted || this.closed) {
					streamQueue.close();
					return;
				}
				streamQueue.close(error);
			}
		};
		startStream();
		const cleanup = () => {
			this.eventStreams.delete(ac);
			ac.abort();
			streamQueue.close();
		};
		return {
			events: { [Symbol.asyncIterator]: () => ({
				next: async () => await streamQueue.shift(),
				return: async () => {
					cleanup();
					return {
						done: true,
						value: void 0
					};
				}
			}) },
			ready,
			close: cleanup
		};
	}
	async close() {
		if (this.closed) return;
		this.closed = true;
		this.sessionAbortController.abort();
		for (const ac of this.eventStreams) ac.abort();
		this.eventStreams.clear();
		this.queue.close();
	}
	async request(path, init) {
		const url = require_utils.toAbsoluteUrl(this.apiUrl, path);
		let requestInit = {
			...init,
			headers: require_utils.mergeHeaders(this.defaultHeaders, init.headers)
		};
		if (this.onRequest) requestInit = await this.onRequest(url, requestInit);
		try {
			const response = await (await this.resolveFetch())(url.toString(), requestInit);
			if (!response.ok) {
				let detail = "";
				try {
					const body = await response.text();
					const parsed = JSON.parse(body);
					if (typeof parsed === "object" && parsed != null) detail = parsed.message ?? parsed.error ?? "";
					if (!detail) detail = body;
				} catch {}
				const message = detail ? `Protocol request failed: ${response.status} ${response.statusText} — ${detail}` : `Protocol request failed: ${response.status} ${response.statusText}`;
				throw new Error(message);
			}
			return response;
		} catch (error) {
			throw require_utils.toError(error);
		}
	}
};
//#endregion
exports.ProtocolSseTransportAdapter = ProtocolSseTransportAdapter;
