// @langchain/langgraph-sdk/dist/client/stream/transport/utils.cjs
//#region src/client/stream/transport/utils.ts
const isRecord = (value) => typeof value === "object" && value !== null;
const toAbsoluteUrl = (apiUrl, path) => new URL(path, apiUrl.endsWith("/") ? apiUrl : `${apiUrl}/`);
const toError = (error) => error instanceof Error ? error : new Error(String(error));
const toWebSocketUrl = (apiUrl) => {
	const url = new URL(apiUrl);
	url.protocol = url.protocol === "https:" ? "wss:" : "ws:";
	url.search = "";
	url.hash = "";
	return url.toString();
};
const hasHeaders = (headers) => Object.values(headers ?? {}).some((value) => value != null);
function mergeHeaders(...headerGroups) {
	const merged = new Headers();
	for (const group of headerGroups) {
		if (!group) continue;
		if (group instanceof Headers) {
			group.forEach((value, key) => {
				merged.set(key, value);
			});
			continue;
		}
		if (Array.isArray(group)) {
			for (const [key, value] of group) if (value == null) merged.delete(key);
			else merged.set(key, value);
			continue;
		}
		for (const [key, value] of Object.entries(group)) if (value == null) merged.delete(key);
		else merged.set(key, value);
	}
	return merged;
}
function isProtocolResponse(value) {
	return isRecord(value) && typeof value.type === "string" && (value.type === "success" || value.type === "error");
}
//#endregion
exports.hasHeaders = hasHeaders;
exports.isProtocolResponse = isProtocolResponse;
exports.isRecord = isRecord;
exports.mergeHeaders = mergeHeaders;
exports.toAbsoluteUrl = toAbsoluteUrl;
exports.toError = toError;
exports.toWebSocketUrl = toWebSocketUrl;
