// @langchain/langgraph-sdk/dist/client/stream/transport/websocket.cjs
const require_queue = require("./queue.cjs");
const require_utils = require("./utils.cjs");
//#region src/client/stream/transport/websocket.ts
/**
* Transport adapter that speaks the thread-centric protocol over a
* bidirectional WebSocket. Bound to a specific `threadId` — the socket
* connects to `ws://.../threads/:thread_id/stream/events`.
*/
var ProtocolWebSocketTransportAdapter = class {
	threadId;
	queue = new require_queue.AsyncQueue();
	apiUrl;
	defaultHeaders;
	onRequest;
	webSocketFactory;
	streamUrl;
	pending = /* @__PURE__ */ new Map();
	socket = null;
	closed = false;
	intentionalClose = false;
	constructor(options) {
		this.apiUrl = options.apiUrl;
		this.threadId = options.threadId;
		this.defaultHeaders = options.defaultHeaders;
		this.onRequest = options.onRequest;
		this.webSocketFactory = options.webSocketFactory ?? ((url) => new WebSocket(url));
		this.streamUrl = options.paths?.stream ?? `/threads/${this.threadId}/stream/events`;
	}
	async open() {
		if (this.socket != null) return;
		this.assertBrowserSafeTransportConfig();
		const wsUrl = require_utils.toWebSocketUrl(new URL(this.streamUrl, this.apiUrl.endsWith("/") ? this.apiUrl : `${this.apiUrl}/`).toString());
		const socket = this.webSocketFactory(wsUrl);
		this.socket = socket;
		this.closed = false;
		this.intentionalClose = false;
		socket.addEventListener("message", this.handleMessage);
		socket.addEventListener("close", this.handleClose);
		socket.addEventListener("error", this.handleSocketError);
		await new Promise((resolve, reject) => {
			const onOpen = () => {
				cleanup();
				resolve();
			};
			const onError = () => {
				cleanup();
				reject(/* @__PURE__ */ new Error("Failed to open protocol WebSocket."));
			};
			const cleanup = () => {
				socket.removeEventListener("open", onOpen);
				socket.removeEventListener("error", onError);
			};
			socket.addEventListener("open", onOpen, { once: true });
			socket.addEventListener("error", onError, { once: true });
		});
	}
	async send(command) {
		return await this.sendCommand(command);
	}
	events() {
		const queue = this.queue;
		return { [Symbol.asyncIterator]: () => ({
			next: async () => await queue.shift(),
			return: async () => {
				queue.close();
				return {
					done: true,
					value: void 0
				};
			}
		}) };
	}
	async close() {
		if (this.closed) return;
		this.closed = true;
		this.intentionalClose = true;
		for (const { reject } of this.pending.values()) reject(/* @__PURE__ */ new Error("Protocol WebSocket connection closed."));
		this.pending.clear();
		this.queue.close();
		const socket = this.socket;
		this.socket = null;
		if (!socket) return;
		await new Promise((resolve) => {
			if (socket.readyState === WebSocket.CLOSED) {
				resolve();
				return;
			}
			const onClose = () => {
				socket.removeEventListener("close", onClose);
				resolve();
			};
			socket.addEventListener("close", onClose, { once: true });
			if (socket.readyState === WebSocket.OPEN || socket.readyState === WebSocket.CONNECTING) socket.close();
			else resolve();
		});
	}
	assertBrowserSafeTransportConfig() {
		if (require_utils.hasHeaders(this.defaultHeaders) || this.onRequest != null) throw new Error("Browser WebSocket protocol transport does not support defaultHeaders or onRequest hooks. Supply a custom protocolWebSocketFactory if you need custom WebSocket setup.");
	}
	async sendCommand(command) {
		const socket = this.socket;
		if (socket == null || socket.readyState !== WebSocket.OPEN) throw new Error("Protocol WebSocket is not open.");
		return await new Promise((resolve, reject) => {
			this.pending.set(command.id, {
				resolve,
				reject
			});
			try {
				socket.send(JSON.stringify(command));
			} catch (error) {
				this.pending.delete(command.id);
				reject(require_utils.toError(error));
			}
		});
	}
	handleMessage = (event) => {
		let payload;
		try {
			payload = JSON.parse(String(event.data));
		} catch {
			return;
		}
		if (require_utils.isRecord(payload) && typeof payload.id === "number" && (payload.type === "success" || payload.type === "error")) {
			const pending = this.pending.get(payload.id);
			if (pending) {
				this.pending.delete(payload.id);
				pending.resolve(payload);
			}
			return;
		}
		if (require_utils.isRecord(payload) && payload.type === "event") this.queue.push(payload);
	};
	handleClose = () => {
		this.socket = null;
		if (this.intentionalClose || this.closed) {
			this.queue.close();
			return;
		}
		const error = /* @__PURE__ */ new Error("Protocol WebSocket closed unexpectedly.");
		for (const { reject } of this.pending.values()) reject(error);
		this.pending.clear();
		this.queue.close(error);
	};
	handleSocketError = () => {
		if (this.closed || this.intentionalClose) return;
		const error = /* @__PURE__ */ new Error("Protocol WebSocket encountered an error.");
		for (const { reject } of this.pending.values()) reject(error);
		this.pending.clear();
		this.queue.close(error);
	};
};
//#endregion
exports.ProtocolWebSocketTransportAdapter = ProtocolWebSocketTransportAdapter;
