// @langchain/langgraph-sdk/dist/stream/root-message-projection.js
import { MessageAssembler } from "../client/stream/messages.js";
import { namespaceKey } from "./namespace.js";
import { assembledMessageToBaseMessage } from "./assembled-to-message.js";
import { buildMessageIndex, messagesEqual, reconcileMessagesFromValues, shouldPreferValuesMessageForToolCalls } from "./message-reconciliation.js";
//#region src/stream/root-message-projection.ts
/**
* Root-namespace message projection. Owns the merge between the
* `messages` (streamed deltas) and `values` (authoritative
* snapshots) channels for the root namespace.
*
* @typeParam StateType     - Root state shape; the messages array is read
*   from `values[messagesKey]`.
* @typeParam InterruptType - Shape of root protocol interrupts (forwarded
*   into `RootSnapshot` updates).
*/
var RootMessageProjection = class {
	/**
	* Key inside `values` that holds the message array. Defaults to
	* `"messages"` in the controller; configurable for state graphs
	* that surface messages under a different slot.
	*/
	#messagesKey;
	/** Root snapshot store written to on every merge. */
	#store;
	/**
	* Subagent discovery runner notified about every assembled root
	* message. Driving discovery from assembled messages (rather than
	* raw events) lets us discover subagents from synthesized
	* `tool_calls` without re-parsing protocol payloads.
	*/
	#subagents;
	/**
	* Stateful chunk assembler for in-flight messages. Reset (via a
	* fresh instance) on every {@link reset} so a new thread starts
	* with no half-built messages from the previous one.
	*/
	#assembler = new MessageAssembler();
	/**
	* `messageId → role/toolCallId` captured from `message-start` events.
	* The assembler's intermediate output drops these fields, so we cache
	* them at start-time and reapply when projecting to a `BaseMessage`.
	*/
	#roles = /* @__PURE__ */ new Map();
	/**
	* `messageId → position in #store.messages` for fast in-place
	* updates as deltas arrive. Rebuilt on every full reconciliation
	* driven by a `values` event.
	*/
	#indexById = /* @__PURE__ */ new Map();
	/**
	* Ids observed in the most recent `values.messages` snapshot.
	* Reconciliation uses this to detect server-side removals: a
	* previously-seen id missing from the next snapshot means it was
	* removed by the server (and should drop from the projection).
	*/
	#valuesMessageIds = /* @__PURE__ */ new Set();
	/**
	* `namespaceKey → tool_call_id` captured from root `tool-started`
	* events. Used as a fallback when a tool-role `message-start` is
	* missing its `tool_call_id` field.
	*/
	#toolCallIdByNamespace = /* @__PURE__ */ new Map();
	/**
	* @param params.messagesKey - Key inside `values` that holds the
	*   message array.
	* @param params.store       - Root snapshot store to mutate.
	* @param params.subagents   - Discovery runner fed by each new
	*   assembled message.
	*/
	constructor(params) {
		this.#messagesKey = params.messagesKey;
		this.#store = params.store;
		this.#subagents = params.subagents;
	}
	/**
	* Drop all per-thread state. Called by the controller on thread
	* rebind / dispose so a swap doesn't surface stale messages.
	*/
	reset() {
		this.#assembler = new MessageAssembler();
		this.#roles.clear();
		this.#indexById.clear();
		this.#valuesMessageIds = /* @__PURE__ */ new Set();
		this.#toolCallIdByNamespace.clear();
	}
	/**
	* Record a `namespace → tool_call_id` mapping captured from a root
	* `tool-started` event.
	*
	* The companion tool-role `message-start` event may not carry a
	* `tool_call_id`, so we fall back to the most recent value recorded
	* here for the same namespace.
	*
	* @param namespace  - Event namespace from the `tool-started` event.
	* @param toolCallId - Tool call id from the same event.
	*/
	recordToolCallNamespace(namespace, toolCallId) {
		this.#toolCallIdByNamespace.set(namespaceKey(namespace), toolCallId);
	}
	/**
	* Apply a `messages` channel event to the projection.
	*
	* Captures role/tool metadata on `message-start`, feeds the chunk
	* to the assembler, projects the assembled output to a
	* {@link BaseMessage}, and either appends or in-place updates the
	* store's messages array based on whether the id was seen before.
	*
	* @param event - The `messages` channel event to consume.
	*/
	handleMessage(event) {
		const data = event.params.data;
		if (data.event === "message-start") {
			const startData = data;
			const role = startData.role ?? "ai";
			const extendedRole = startData.role ?? role;
			let toolCallId = startData.tool_call_id;
			if (extendedRole === "tool" && toolCallId == null) {
				const messageId = startData.id;
				if (messageId != null) {
					const match = /-tool-(.+)$/.exec(messageId);
					if (match != null) toolCallId = match[1];
				}
				if (toolCallId == null) toolCallId = this.#toolCallIdByNamespace.get(namespaceKey(event.params.namespace));
			}
			if (startData.id != null) this.#roles.set(startData.id, {
				role: extendedRole,
				toolCallId
			});
		}
		const update = this.#assembler.consume(event);
		if (update == null) return;
		const id = update.message.id;
		if (id == null) return;
		const captured = this.#roles.get(id) ?? { role: "ai" };
		const base = assembledMessageToBaseMessage(update.message, captured.role, { toolCallId: captured.toolCallId });
		this.#subagents.discoverFromMessage(base, event.params.namespace);
		this.#store.setState((s) => {
			const existingIdx = this.#indexById.get(id);
			let messages;
			if (existingIdx == null) {
				this.#indexById.set(id, s.messages.length);
				messages = [...s.messages, base];
			} else if (messagesEqual(s.messages[existingIdx], base)) return s;
			else {
				messages = s.messages.slice();
				messages[existingIdx] = base;
			}
			const values = syncMessagesIntoValues(s.values, this.#messagesKey, messages);
			return values === s.values ? {
				...s,
				messages
			} : {
				...s,
				messages,
				values
			};
		});
	}
	/**
	* Reconcile a full `values` snapshot into the projection.
	*
	* Delegates the merge to {@link reconcileMessagesFromValues}:
	* values stays authoritative for ordering and removals, while
	* streamed in-flight messages keep their content until the server
	* echoes them back. Empty messages just refresh the values blob.
	*
	* Rebuilds {@link #indexById} after the merge so subsequent delta
	* applications target the new positions.
	*
	* @param nextValues   - Full values snapshot from the `values` event.
	* @param nextMessages - The messages array extracted from
	*   `values[messagesKey]` and coerced to `BaseMessage` instances.
	*/
	applyValues(nextValues, nextMessages) {
		this.#store.setState((s) => {
			if (nextMessages.length === 0) return stateValuesShallowEqual(s.values, nextValues, this.#messagesKey) ? s : {
				...s,
				values: nextValues
			};
			const reconciliation = reconcileMessagesFromValues({
				valueMessages: nextMessages,
				currentMessages: s.messages,
				currentIndexById: this.#indexById,
				previousValueMessageIds: this.#valuesMessageIds,
				preferValuesMessage: shouldPreferValuesMessageForToolCalls
			});
			this.#valuesMessageIds = reconciliation.valueMessageIds;
			const messages = reconciliation.messages;
			const values = {
				...nextValues,
				[this.#messagesKey]: messages
			};
			if (messages === s.messages && stateValuesShallowEqual(s.values, values, this.#messagesKey)) return s;
			this.#indexById.clear();
			for (const [id, idx] of buildMessageIndex(messages)) this.#indexById.set(id, idx);
			return {
				...s,
				values,
				messages
			};
		});
	}
};
/**
* Mirror a freshly-updated message list into `values[messagesKey]`.
*
* Returns the same `values` reference when the list is already
* equal-by-content so the caller can keep the existing snapshot
* identity (and avoid spurious `setSnapshot` notifications).
*/
function syncMessagesIntoValues(values, messagesKey, messages) {
	const record = values;
	const current = record[messagesKey];
	if (Array.isArray(current) && messagesEqualList(current, messages)) return values;
	return {
		...record,
		[messagesKey]: messages
	};
}
/**
* True when two `BaseMessage` arrays carry the same per-message
* content (using {@link messagesEqual}).
*/
function messagesEqualList(previous, next) {
	if (previous === next) return true;
	if (previous.length !== next.length) return false;
	for (let i = 0; i < previous.length; i += 1) if (!messagesEqual(previous[i], next[i])) return false;
	return true;
}
/**
* Shallow-equal for `values` objects, *ignoring* the messages slot.
*
* The messages array is compared separately by the caller (via
* {@link messagesEqualList}) because both arrays contain class
* instances whose JSON representation is not stable across reads.
*/
function stateValuesShallowEqual(previous, next, messagesKey) {
	if (previous === next) return true;
	const previousRecord = previous;
	const nextRecord = next;
	const previousKeys = Object.keys(previousRecord);
	const nextKeys = Object.keys(nextRecord);
	if (previousKeys.length !== nextKeys.length) return false;
	for (const key of previousKeys) {
		if (!Object.prototype.hasOwnProperty.call(nextRecord, key)) return false;
		const previousValue = previousRecord[key];
		const nextValue = nextRecord[key];
		if (key === messagesKey && Array.isArray(previousValue) && Array.isArray(nextValue)) continue;
		if (!Object.is(previousValue, nextValue)) return false;
	}
	return true;
}
//#endregion
export { RootMessageProjection };
